UIforETW/Utility.cpp

```diff
@@ -594,69 +871,171 @@ void SetCurrentThreadName(PCSTR threadName)
+}
+
+void CloseRegKey(_In_ _Pre_valid_ _Post_ptr_invalid_ const HKEY key, PCWSTR const keyName)
+{
+	const LONG closeKey = ::RegCloseKey(key);
+	if (closeKey == ERROR_SUCCESS)
+		return;
+	outputPrintf(L"Failed to close registry key `%s`.\n", keyName);
+	debugLastError();
+}
+
+void ClipboardClose()
+{
+	const BOOL closeResult = ::CloseClipboard( );
+	if (closeResult)
+		return;
+	debugPrintf(L"Failed to close the clipboard!\n");
+	debugLastError( );
+}
+
+bool ClipboardOpen()
+{
+	const BOOL openClip = ::OpenClipboard(::GetDesktopWindow());
+	if (!openClip)
 	{
-		std::wstring source = exeDir + fileName;
-		std::wstring destDir = documents + std::wstring(L"\\WPA Files");
-		std::wstring dest = destDir + fileName;
-		if (force || !PathFileExists(dest.c_str()))
-		{
-			(void)_wmkdir(destDir.c_str());
-			if (CopyFile(source.c_str(), dest.c_str(), FALSE))
-			{
-				if (force)
-					outputPrintf(L"Copied Startup.wpaProfile to the WPA Files directory.\n");
-			}
-			else
-			{
-				if (force)
-					outputPrintf(L"Failed to copy Startup.wpaProfile to the WPA Files directory.\n");
-			}
-		}
+		outputPrintf(L"Failed to open clipboard!\n");
+		debugLastError();
+		return false;
 	}
+	return true;
+}
 
-	// Then copy the WPA 10 startup.wpaProfile file
-#pragma warning(suppress : 4996)
-	const wchar_t* localAppData = _wgetenv(L"localappdata");
-	if (localAppData)
-	{
-		std::wstring source = exeDir + L"\\startup10.wpaProfile";
-		std::wstring destDir = std::wstring(localAppData) + L"\\Windows Performance Analyzer";
-		std::wstring dest = destDir + fileName;
-		if (force || !PathFileExists(dest.c_str()))
-		{
-			(void)_wmkdir(destDir.c_str());
-			if (CopyFile(source.c_str(), dest.c_str(), FALSE))
-			{
-				if (force)
-					outputPrintf(L"%s", L"Copied Startup.10wpaProfile to %localappdata%\\Windows Performance Analyzer\n");
-			}
-			else
-			{
-				if (force)
-					outputPrintf(L"%s", L"Failed to copy Startup.10wpaProfile to %localappdata%\\Windows Performance Analyzer\n");
-			}
-		}
+bool ClipboardEmpty()
+{
+	const BOOL emptyClip = ::EmptyClipboard();
+	if (!emptyClip)
+	{
+		outputPrintf(L"Failed to empty clipboard!\n");
+		debugLastError();
+		return false;
 	}
+	return true;
 }
 
-void CloseValidHandle( _Pre_valid_ _Post_ptr_invalid_ HANDLE handle )
+_Success_(return)
+bool AllocGlobalMemoryForString(const std::wstring& text, _Out_ _Post_valid_ HGLOBAL* const hmem)
 {
-	const BOOL handleClosed = ::CloseHandle(handle);
-	if (handleClosed == 0)
+	const size_t length = (text.size() + 1) * sizeof(text[0]);
+	const HANDLE mem = ::GlobalAlloc(GMEM_MOVEABLE, length);
+	if (!mem)
 	{
-		std::terminate( );
+		outputPrintf(L"Couldn't allocate memory to set clipboard text!\n");
+		outputLastError();
+		return false;
 	}
+	(*hmem) = mem;
+	return true;
+}
 
+void FreeGlobalMemory(_Frees_ptr_opt_ const HGLOBAL hmem)
+{
+		//If [GlobalFree] succeeds, the return value is NULL.
+		const HGLOBAL freeRes = ::GlobalFree(hmem);
+		if (freeRes != NULL)

```

Comment:
I believe that if `::GlobalFree` fails, it's the result of a serious logic bug? I could instead call `outputLastError` to make the error visible to the user without killing the whole process.